osu.Game/Graphics/ParticleSpewer.cs

```diff
@@ -0,0 +1,208 @@
+
+        public float ParticleGravity;
+
+        private bool hasActiveParticles => Active.Value || (lastParticleAdded + maxDuration) > Time.Current;
+
+        public ParticleSpewer(Texture texture, int perSecond, double maxDuration)
+        {
+            Texture = texture;
+            Blending = BlendingParameters.Additive;
+
+            particles = new FallingParticle[perSecond * (int)Math.Ceiling(maxDuration / 1000)];
+
+            cooldown = 1000f / perSecond;
+            this.maxDuration = maxDuration;
+        }
+
+        protected override void Update()
+        {
+            base.Update();
+
+            // reset cooldown if the clock was rewound.
+            // this can happen when seeking in replays.
+            if (lastParticleAdded > Time.Current) lastParticleAdded = 0;
+
+            if (Active.Value && Time.Current > lastParticleAdded + cooldown)
+            {
+                var newParticle = CreateParticle();
+
+                if (newParticle.HasValue)
+                {
+                    var particle = newParticle.Value;
+                    particle.StartTime = (float)Time.Current;
+
+                    particles[currentIndex] = particle;
+
+                    currentIndex = (currentIndex + 1) % particles.Length;
+                    lastParticleAdded = Time.Current;
+                }
+            }
+
+            Invalidate(Invalidation.DrawNode);
+        }
+
+        protected override DrawNode CreateDrawNode() => new ParticleSpewerDrawNode(this);
+
+        # region DrawNode
+
+        private class ParticleSpewerDrawNode : SpriteDrawNode
+        {
+            private readonly FallingParticle[] particles;
+
+            protected new ParticleSpewer Source => (ParticleSpewer)base.Source;
+
+            private readonly float maxDuration;
+
+            private float currentTime;
+            private float gravity;
+            private Axes relativePositionAxes;
+            private Vector2 sourceSize;
+
+            public ParticleSpewerDrawNode(Sprite source)
+                : base(source)
+            {
+                particles = new FallingParticle[Source.particles.Length];
+                maxDuration = (float)Source.maxDuration;
+            }
+
+            public override void ApplyState()
+            {
+                base.ApplyState();
+
+                Source.particles.CopyTo(particles, 0);
+
+                currentTime = (float)Source.Time.Current;
+                gravity = Source.ParticleGravity;
+                relativePositionAxes = Source.RelativePositionAxes;
+                sourceSize = Source.DrawSize;
+            }
+
+            protected override void Blit(Action<TexturedVertex2D> vertexAction)
+            {
+                foreach (var p in particles)
+                {
+                    // ignore particles that weren't initialized.
+                    if (p.StartTime <= 0) continue;
```

Comment:
This isn't 100% right. The start time can be earlier than 0 and still not mean that the particle wasn't initialised. That will be the case with beatmaps with a long lead-in time / long storyboarded intro, and in those cases the gameplay clock does not start at time 0.

An example map that this can be seen on is https://osu.ppy.sh/beatmapsets/5381#osu/26567 (stable still shows the particles even in the intro section).

Reply:
Should be fixed now. I also added a test for these cases.